### PR TITLE
[staging] Improve c-ares packaging

### PR DESCRIPTION
**Pull request by jonringer**
###### Description of changes

Improved version of https://github.com/NixOS/nixpkgs/pull/191120 and https://github.com/NixOS/nixpkgs/pull/191123

expose `c-aresMinimal` for curl + nghttp2 to be used for bootstrapping + `fetchpatch`.

But expose `c-ares` with cmake build to everyone else. This is a major improvement to people consuming this through CMake.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
@GrahamcOfBorg build grpc

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Review by tobim — APPROVED** [review]
Thanks for getting rid of this custom hack!

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Comment by mweinelt** [comment]
@ofborg build c-ares

**Comment by mweinelt** [comment]
c-aresMinimal

```
#***************************************************************************
# Project        ___       __ _ _ __ ___  ___ 
#               / __|____ / _` | '__/ _ \/ __|
#              | (_|_____| (_| | | |  __/\__ \
#               \___|     \__,_|_|  \___||___/
#
prefix=/nix/store/h82dm9xkxhsirnbsgnaayx3ghc9bvzbi-c-ares-1.18.1
exec_prefix=${prefix}
libdir=/nix/store/h82dm9xkxhsirnbsgnaayx3ghc9bvzbi-c-ares-1.18.1/lib
includedir=/nix/store/6y1n5rja7q5fjzgvqknckgs5x8fz4c6b-c-ares-1.18.1-dev/include

Name: c-ares
URL: http://c-ares.org/
Description: asynchronous DNS lookup library
Version: 1.18.1
Requires: 
Requires.private: 
Cflags: -I${includedir} 
Libs: -L${libdir} -lcares
Libs.private: 
```

c-ares

```
#***************************************************************************
# Project        ___       __ _ _ __ ___  ___ 
#               / __|____ / _` | '__/ _ \/ __|
#              | (_|_____| (_| | | |  __/\__ \
#               \___|     \__,_|_|  \___||___/
#
prefix=/nix/store/6qjnsagabw6vf2vqvp6blc5rd6661kdx-c-ares-1.18.1
exec_prefix=${prefix}//nix/store/6qjnsagabw6vf2vqvp6blc5rd6661kdx-c-ares-1.18.1/bin
libdir=${prefix}//nix/store/6qjnsagabw6vf2vqvp6blc5rd6661kdx-c-ares-1.18.1/lib
includedir=/nix/store/pgvk94m98y2sq5x91mpz2nsm7pd2yqkp-c-ares-1.18.1-dev//nix/store/pgvk94m98y2sq5x91mpz2nsm7pd2yqkp-c-ares-1.18.1-dev/include

Name: c-ares
URL: https://c-ares.org/
Description: asynchronous DNS lookup library
Version: 1.18.1
Requires: 
Requires.private: 
Cflags: -I${includedir} 
Libs: -L${libdir} -lcares
Libs.private: 
```

**Comment by Artturin** [comment]
c-aresMinimal pc.in https://github.com/c-ares/c-ares/blob/main/libcares.pc.in

c-ares pc.in https://github.com/c-ares/c-ares/blob/main/libcares.pc.cmake

**Comment by jonringer** [comment]
ugh, why can't anyone use CMake correctly....

(It's just a joke, it's almost impossible to find out what you're supposed to do with CMake as everything is 20 years of hacks around trying to do the right thing)

**Comment by jonringer** [comment]
Upstream PR: https://github.com/c-ares/c-ares/pull/486

**Comment by jonringer** [comment]
```
[13:01:24] jon@nixos ~/projects/nixpkgs (improve-c-ares)
$ nix-build -A c-ares.tests
/nix/store/wr43s9xx96ak7v7m04n6b7qamd2q0glp-curl-7.85.0-bin
/nix/store/8kgh27mkhq6lyngwmyggri0lv53n6n4k-grpc-1.48.1
```

**Comment by jonringer** [comment]
darwin errors look related to bootstrapping failing to build

**Comment by mweinelt** [comment]
Yeah, somewhere in libgmp :shrug: I think this looks good and is a huge improvement. Thanks for working on it.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -18032,7 +18032,13 @@ with pkgs;
 
   bzrtp = callPackage ../development/libraries/bzrtp { };
 
-  c-ares = callPackage ../development/libraries/c-ares { };
+  c-ares = callPackage ../development/libraries/c-ares {
+    inherit (buildPackages) cmake;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18036]: We should do this more often to fix cross builds.

Comment (Artturin) [reply pkgs/top-level/all-packages.nix:18036]: actually we should just not pass packages with callPackage

also i don't see a reason why cmake should be inherit here

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:18036]: Oh, in the context of python-packages this would make sense but not all-packages.

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:18036]: #196038

Comment (jonringer) [reply pkgs/top-level/all-packages.nix:18036]: yea, cmake should be getting added correctly as it's listed as a `nativeBuildInputs`. But i dont' see much harm in it either

**Pull request merged**